src/olympia/devhub/views.py

```diff
@@ -1387,109 +1390,207 @@ def version_stats(request, addon_id, addon):
-                  {'distribution_form': form})
+                  {'distribution_form': form,
+                   'submit': 'version' if addon else 'addon'})
 
 
 @login_required
+def submit_addon_distribution(request):
+    return _submit_distribution(request, None, 'devhub.submit.upload')
+
+
+@dev_required
+@waffle_switch('step-version-upload')
+@waffle_switch('mixed-listed-unlisted')
+def submit_version_distribution(request, addon_id, addon):
+    return _submit_distribution(request, addon, 'devhub.submit.version.upload')
+
+
 @transaction.atomic
-def submit_addon_upload(request, channel):
-    form = forms.NewAddonForm(
+def _submit_upload(request, addon, channel, next_listed, next_unlisted):
+    form = forms.NewVersionForm(
         request.POST or None,
         request.FILES or None,
+        addon=addon,
         request=request
     )
-    is_listed = channel == 'listed'
-    if request.method == 'POST':
-        if form.is_valid():
-            data = form.cleaned_data
-
-            p = data.get('supported_platforms', [])
+    if request.method == 'POST'and form.is_valid():
+        data = form.cleaned_data
+        is_beta = (data['beta'] and addon and
+                   channel == amo.RELEASE_CHANNEL_LISTED)
 
-            addon = Addon.from_upload(data['upload'], p, source=data['source'],
-                                      is_listed=is_listed)
+        if addon:
+            version = Version.from_upload(
+                upload=form.cleaned_data['upload'],
+                addon=addon,
+                platforms=data.get('supported_platforms', []),
+                channel=channel,
+                source=data['source'],
+                is_beta=is_beta)
+            url_args = [addon.slug, version.id]
+        else:
+            addon = Addon.from_upload(
+                upload=data['upload'],
+                platforms=data.get('supported_platforms', []),
+                source=data['source'],
+                is_listed=channel == amo.RELEASE_CHANNEL_LISTED)
+            version = addon.current_version
             AddonUser(addon=addon, user=request.user).save()
-            check_validation_override(request, form, addon,
-                                      addon.current_version)
-            if not addon.is_listed:  # Not listed? Automatically choose queue.
-                addon.update(status=amo.STATUS_NOMINATED)
-                # Sign all the files submitted, one for each platform.
-                auto_sign_version(addon.versions.get())
-                return redirect('devhub.submit.finish', addon.slug)
-            return redirect('devhub.submit.details', addon.slug)
-    is_admin = acl.action_allowed(request, 'ReviewerAdminTools', 'View')
+            url_args = [addon.slug]
 
+        check_validation_override(request, form, addon, version)
+        if data['source']:
+            addon_update = {'admin_review': True}
+        else:
+            addon_update = {}
```

Comment:
Might be slightly easier to read and more future-proof to always do `addon_update = {}` above, then add things to the dict if necessary. Something like:
```python
addon_update = {}
if data['source']:
    addon_update['admin_review'] = True
if addon.status == amo.STATUS_NULL and addon.has_complete_metadata():
    addon_update['status'] = amo.STATUS_NOMINATED

```